storm/Error.hpp

```diff
@@ -9,27 +9,90 @@
 
 #if defined(WHOA_SYSTEM_MAC) || defined(WHOA_SYSTEM_LINUX)
 #define ERROR_SUCCESS           0x0
+#define ERROR_INVALID_HANDLE    0x6
+#define ERROR_NOT_ENOUGH_MEMORY 0x8
 #define ERROR_INVALID_PARAMETER 0x57
 #endif
 
 #if defined(NDEBUG)
-#define STORM_ASSERT(x)                          \
-    (void)0
+#define STORM_ASSERT(x)
+#define STORM_ASSERT_FATAL(x)
 #else
-#define STORM_ASSERT(x)                          \
+#define STORM_ASSERT(x)                                                                     \
+    if (!(x)) {                                                                             \
+        SErrDisplayError(STORM_ERROR_ASSERTION, __FILE__, __LINE__, #x, 0, 1, 0x11111111);  \
+    }
+
```

Comment:
When do we think `STORM_ASSERT` is called? WoW 0.5.3 seems to make heavy use of it based on calls to `SErrDisplayError`, but I'm not positive if 0.5.3 is a debug build or a release assertions enabled build.

Reply:
Alright, went spelunking. Here's what I've found so far:

- `4.1.0.13682` -  `WoW [Release Assertions Enabled] Build 13682 (Feb 22 2011)`
  - Specimen: `SStrHash` @ `0x00154CE0`
    - validation macros only call `SErrDisplayError`
    - no calls to `SErrPrepareAppFatal` or `SErrDisplayAppFatal`
    - no call to `SErrSetLastError`
    - effectively, a non-fatal validation assert that *doesn't* set last error code (intriguing)

- `3.3.5.12340` - `WoW [Release] Build 12340 (Jun 24 2010 23:54:57)`
  - Specimen: `SStrHash` @ `0x0076F340`
    - validation macros only call `SErrSetLastError`
    - no calls to `SErrDisplayError`
    - no calls to `SErrPrepareAppFatal` or `SErrDisplayAppFatal`
    - effectively, validation assertions are suppressed and only last error codes are tracked

- `0.5.3.????` - `WoW [Release Assertions Enabled] Build 3368 (Dec 11 2003 18:01:27)`
  - Specimen: `SStrHash` @ `0x0063A2B0`
    - validation macros call `SErrPrepareAppFatal` and `SErrDisplayAppFatal`
    - validation macros also call `SErrSetLastError`
    - no calls to `SErrDisplayError`
    - effectively, a fatal assert that also sets last error code